NeuroID/src/main/java/com/neuroid/tracker/events/TouchEventManager.kt

```diff
@@ -90,71 +128,198 @@ abstract class TouchEventManager {
+            val etnSenderName = getEtnSenderName(currentView)
+            val gyroData = NIDSensorHelper.getGyroscopeInfo()
+            val accelData = NIDSensorHelper.getAccelerometerInfo()
+
+            detectChangesOnView(currentView, timeMills, motionEvent.action)
+
+            val typeOfView = detectViewType(currentView)
+
+            var v = ""
+            val metadataObj = mutableMapOf<String, Any>()
+
+            when (currentView) {
+                is EditText -> {
+                    v = "S~C~~${currentView.text.length}"
+                }
+                is RadioButton -> {
+                    metadataObj.put("type", "radioButton")
+                    metadataObj.put("id", currentView.getIdOrTag())
+
+                    // go up to 3 parents in case a RadioGroup is not the direct parent
+                    var rParent = currentView.parent
+                    repeat(3) { index ->
+                        if (rParent is RadioGroup) {
+                            val p = rParent as RadioGroup
+                            metadataObj.put("rGroupId", "${p.getIdOrTag()}")
+                            return@repeat
+                        } else {
+                            rParent = rParent.parent
+                        }
+                    }
+                }
             }
-        )
-        metadataObj.put("xP",
-            if(motionEvent?.xPrecision != null) {
-                motionEvent.xPrecision
-            } else {
-                0
+
+            var motionValues = mapOf<String, Any>()
+            try {
+                motionValues = generateMotionEventValues(motionEvent)
+            } catch (ex: Exception) {
+                NIDLog.d("TouchEventManager", "no motion error: ${ex.printStackTrace()}")
             }
-        )
-        metadataObj.put("xR",
-            if(motionEvent?.rawX != null) {
-                motionEvent.rawX
-            } else {
-                0
+
+            val rawAction = mapOf("rawAction" to it.action)
+            val attrJSON = listOf(rawAction, metadataObj, motionValues)
+
+            var shouldSaveEvent = false
+            var eventType = ""
+            when (it.action) {
+                MotionEvent.ACTION_DOWN -> {
+                    if (typeOfView > 0) {
+                        lastViewName = nameView
+                        lastTypeOfView = typeOfView
+
+                        shouldSaveEvent = true
+                        eventType = TOUCH_START
+                    }
+                }
+                MotionEvent.ACTION_MOVE -> {
+                    shouldSaveEvent = true
+                    eventType = TOUCH_START
+                }
+                MotionEvent.ACTION_UP -> {
+                    if (lastTypeOfView > 0) {
+                        lastTypeOfView = 0
+                        lastViewName = ""
+
+                        shouldSaveEvent = true
+                        eventType = TOUCH_START
```

Comment:
This should be eventType = TOUCH_END since this is an ACTION_UP.  This will fix the test09ValidateTouchEnd() test.

Reply:
Oh good catch, I must have missed that and the TOUCH_MOVE, updated now!